Installation trouble

Greetings! I have failures within build setup, which i could't resolve. i have the following configuration: i7 10700 without discrete video, debian system. So i need to run al least on cpu, further on integrated gpu.

I do intallation in this manner:
cmake -DWITH_ROCM_BACKEND=OFF -DLLVM_DIR=../llvmBuild/llvm/lib -DWITH_CUDA_BACKEND=OFF -DWITH_ACCELERATED_CPU=ON -DCMAKE_INSTALL_PREFIX=../install -S ../hipSYCL-0.9.3
make install

and have some troubles after:
 /home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:56:14: error: 'RegisterStandardPasses' in namespace 'llvm' does not name a type
   56 | static llvm::RegisterStandardPasses
      |              ^~~~~~~~~~~~~~~~~~~~~~
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:60:14: error: 'RegisterStandardPasses' in namespace 'llvm' does not name a type
   60 | static llvm::RegisterStandardPasses
      |              ^~~~~~~~~~~~~~~~~~~~~~
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:74:14: error: 'RegisterStandardPasses' in namespace 'llvm' does not name a type
   74 | static llvm::RegisterStandardPasses
      |              ^~~~~~~~~~~~~~~~~~~~~~
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:84:14: error: 'RegisterStandardPasses' in namespace 'llvm' does not name a type
   84 | static llvm::RegisterStandardPasses
      |              ^~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [src/compiler/CMakeFiles/hipSYCL_clang.dir/build.make:76: src/compiler/CMakeFiles/hipSYCL_clang.dir/HipsyclClangPlugin.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:265: src/compiler/CMakeFiles/hipSYCL_clang.dir/all] Error 2
make: *** [Makefile:136: all] Error 2


____________________________________________________________________________________________________________________________

Second approach i did with -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_COMPILER=clang and get little another message,but also connected with llvm namespace:
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:56:14: error: no type named 'RegisterStandardPasses' in namespace 'llvm'
static llvm::RegisterStandardPasses
       ~~~~~~^
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:57:67: error: no member named 'EP_EnabledOnOptLevel0' in 'llvm::PassManagerBuilder'
    RegisterGlobalsPruningPassOptLevel0(llvm::PassManagerBuilder::EP_EnabledOnOptLevel0,
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~^
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:60:14: error: no type named 'RegisterStandardPasses' in namespace 'llvm'
static llvm::RegisterStandardPasses
       ~~~~~~^
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:61:71: error: no member named 'EP_OptimizerLast' in 'llvm::PassManagerBuilder'
    RegisterGlobalsPruningPassOptimizerLast(llvm::PassManagerBuilder::EP_OptimizerLast,
                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~^
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:74:14: error: no type named 'RegisterStandardPasses' in namespace 'llvm'
static llvm::RegisterStandardPasses
       ~~~~~~^
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:75:76: error: no member named 'EP_EarlyAsPossible' in 'llvm::PassManagerBuilder'
    RegisterLoopSplitAtBarrierPassOptimizerFirst(llvm::PassManagerBuilder::EP_EarlyAsPossible,
                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~^
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:84:14: error: no type named 'RegisterStandardPasses' in namespace 'llvm'
static llvm::RegisterStandardPasses
       ~~~~~~^
/home/alexey/code/hipSYCL/hipSYCL-0.9.3/src/compiler/HipsyclClangPlugin.cpp:85:68: error: no member named 'EP_VectorizerStart' in 'llvm::PassManagerBuilder'
    RegisterMarkParallelBeforeVectorizer(llvm::PassManagerBuilder::EP_VectorizerStart,
                                         ~~~~~~~~~~~~~~~~~~~~~~~~~~^
8 errors generated.
make[2]: *** [src/compiler/CMakeFiles/hipSYCL_clang.dir/build.make:76: src/compiler/CMakeFiles/hipSYCL_clang.dir/HipsyclClangPlugin.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:265: src/compiler/CMakeFiles/hipSYCL_clang.dir/all] Error 2


i see that problem in llvm, but i can't understand it deeply. By far, i tried to install llvm from your script at install directory.

> Hi, you seem to be using a pretty new LLVM that has some legacy infrastructure removed. We adapted hipSYCL to deal with that in a commit post 0.9.3, so could you try if the develop branch works for you?

do you mean to download develop version? Yes , i did. It has changed prompt a little:
In file included from /home/alexey/code/hipSYCL/hipSYCL-develop/src/compiler/cbs/LoopSplitterInlining.cpp:31:
/home/alexey/code/hipSYCL/hipSYCL-develop/src/compiler/../../include/hipSYCL/compiler/cbs/SplitterAnnotationAnalysis.hpp:71:9: error: 'Optional' in namespace 'llvm' does not name a template type
   71 |   llvm::Optional<SplitterAnnotationInfo> SplitterAnnotation_;
      |         ^~~~~~~~
/home/alexey/code/hipSYCL/hipSYCL-develop/src/compiler/../../include/hipSYCL/compiler/cbs/SplitterAnnotationAnalysis.hpp: In member function 'const hipsycl::compiler::SplitterAnnotationInfo& hipsycl::compiler::SplitterAnnotationAnalysisLegacy::getAnnotationInfo() const':
/home/alexey/code/hipSYCL/hipSYCL-develop/src/compiler/../../include/hipSYCL/compiler/cbs/SplitterAnnotationAnalysis.hpp:84:69: error: 'SplitterAnnotation_' was not declared in this scope; did you mean 'SplitterAnnotationInfo'?
   84 |   const SplitterAnnotationInfo &getAnnotationInfo() const { return *SplitterAnnotation_; }
      |                                                                     ^~~~~~~~~~~~~~~~~~~
      |                                                                     SplitterAnnotationInfo
/home/alexey/code/hipSYCL/hipSYCL-develop/src/compiler/../../include/hipSYCL/compiler/cbs/SplitterAnnotationAnalysis.hpp: In member function 'hipsycl::compiler::SplitterAnnotationInfo& hipsycl::compiler::SplitterAnnotationAnalysisLegacy::getAnnotationInfo()':
/home/alexey/code/hipSYCL/hipSYCL-develop/src/compiler/../../include/hipSYCL/compiler/cbs/SplitterAnnotationAnalysis.hpp:85:57: error: 'SplitterAnnotation_' was not declared in this scope; did you mean 'SplitterAnnotationInfo'?
   85 |   SplitterAnnotationInfo &getAnnotationInfo() { return *SplitterAnnotation_; }
      |                                                         ^~~~~~~~~~~~~~~~~~~
      |                                                         SplitterAnnotationInfo
make[2]: *** [src/compiler/CMakeFiles/hipSYCL_clang.dir/build.make:104: src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/LoopSplitterInlining.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:410: src/compiler/CMakeFiles/hipSYCL_clang.dir/all] Error 2
make: *** [Makefile:136: all] Error 2

__________________________________________________________________________________________________-
and with -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++:
In file included from /home/alexey/code/hipSYCL/hipSYCL-develop/src/compiler/cbs/LoopSplitterInlining.cpp:31:
/home/alexey/code/hipSYCL/hipSYCL-develop/src/compiler/../../include/hipSYCL/compiler/cbs/SplitterAnnotationAnalysis.hpp:71:3: error: no template named 'Optional' in namespace 'llvm'; did you mean 'std::optional'?
  llvm::Optional<SplitterAnnotationInfo> SplitterAnnotation_;
  ^~~~~~
/usr/lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/optional:659:11: note: 'std::optional' declared here
    class optional
          ^
1 error generated.
make[2]: *** [src/compiler/CMakeFiles/hipSYCL_clang.dir/build.make:104: src/compiler/CMakeFiles/hipSYCL_clang.dir/cbs/LoopSplitterInlining.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:410: src/compiler/CMakeFiles/hipSYCL_clang.dir/all] Error 2
make: *** [Makefile:136: all] Error 2

--- a/include/hipSYCL/compiler/cbs/SplitterAnnotationAnalysis.hpp
+++ b/include/hipSYCL/compiler/cbs/SplitterAnnotationAnalysis.hpp
@@ -32,6 +32,8 @@
 #include <llvm/IR/LegacyPassManagers.h>
 #include <llvm/IR/PassManager.h>
 
+#include <optional>
+
 namespace hipsycl {
 namespace compiler {
 
@@ -68,7 +70,7 @@ public:
  *       As the annotations should not change from call to call, we cache the result in an Optional.
  */
 class SplitterAnnotationAnalysisLegacy : public llvm::FunctionPass {
-  llvm::Optional<SplitterAnnotationInfo> SplitterAnnotation_;
+  std::optional<SplitterAnnotationInfo> SplitterAnnotation_;
 
 public:
   static char ID;

--- a/src/compiler/cbs/PHIsToAllocas.cpp
+++ b/src/compiler/cbs/PHIsToAllocas.cpp
@@ -74,9 +74,7 @@ llvm::Instruction *breakPHIToAllocas(llvm::PHINode *Phi) {
 bool demotePHIsToAllocas(llvm::Function &F) {
   std::vector<llvm::PHINode *> PHIs;
 
-  auto &BBsInWI = F.getBasicBlockList();
-
-  for (auto &BB : BBsInWI)
+  for (auto &BB : F)
     for (auto &I : BB)
       if (auto *PHI = llvm::dyn_cast<llvm::PHINode>(&I))
         PHIs.push_back(PHI);

--- a/src/compiler/llvm-to-backend/AddressSpaceInferencePass.cpp
+++ b/src/compiler/llvm-to-backend/AddressSpaceInferencePass.cpp
@@ -131,9 +131,9 @@ llvm::PreservedAnalyses AddressSpaceInferencePass::run(llvm::Module &M,
   // need to fix this now.
   unsigned AllocaAddrSpace = ASMap[AddressSpace::AllocaDefault];
   llvm::SmallVector<llvm::Instruction*, 16> InstsToRemove;
-  for(auto& F : M.getFunctionList()) {
-    for(auto& BB : F.getBasicBlockList()) {
-      for(auto& I : BB.getInstList()) {
+  for(auto& F : M) {
+    for(auto& BB : F) {
+      for(auto& I : BB) {
         if(auto* AI = llvm::dyn_cast<llvm::AllocaInst>(&I)) {
           if(AI->getAddressSpace() != AllocaAddrSpace) {
             HIPSYCL_DEBUG_INFO << "AddressSpaceInferencePass: Found alloca in address space "

--- a/src/compiler/llvm-to-backend/LLVMToBackend.cpp
+++ b/src/compiler/llvm-to-backend/LLVMToBackend.cpp
@@ -183,11 +183,11 @@ bool LLVMToBackendTranslator::prepareIR(llvm::Module &M) {
     // Before optimizing, make sure everything has internal linkage to
     // help inlining. All linking should have occured by now, except
     // for backend builtin libraries like libdevice etc
-    for(auto & F : M.getFunctionList()) {
+    for(auto & F : M) {
       // Ignore kernels and intrinsics
       if(!F.isIntrinsic() && !this->isKernelAfterFlavoring(F)) {
         // Ignore undefined functions
-        if(!F.getBasicBlockList().empty())
+        if(!F.empty())
           F.setLinkage(llvm::GlobalValue::InternalLinkage);
       }
     }

--- a/src/compiler/llvm-to-backend/amdgpu/LLVMToAmdgpu.cpp
+++ b/src/compiler/llvm-to-backend/amdgpu/LLVMToAmdgpu.cpp
@@ -145,9 +145,9 @@ bool LLVMToAmdgpuTranslator::toBackendFlavor(llvm::Module &M, PassHandler& PH) {
   // amdgpu does not like some function calls, so try to inline
   // everything. Note: This should be done after ASI pass has fixed
   // alloca address spaces, in case alloca values are passed as arguments!
-  for(auto& F: M.getFunctionList()) {
+  for(auto& F: M) {
     if(F.getCallingConv() != llvm::CallingConv::AMDGPU_KERNEL) {
-      if(!F.getBasicBlockList().empty()) {
+      if(!F.empty()) {
         F.addFnAttr(llvm::Attribute::AlwaysInline);
       }
     }

--- a/src/compiler/llvm-to-backend/spirv/LLVMToSpirv.cpp
+++ b/src/compiler/llvm-to-backend/spirv/LLVMToSpirv.cpp
@@ -126,7 +126,7 @@ bool LLVMToSpirvTranslator::toBackendFlavor(llvm::Module &M, PassHandler& PH) {
     }
   }
 
-  for(auto& F : M.getFunctionList()) {
+  for(auto& F : M) {
     if(F.getCallingConv() != llvm::CallingConv::SPIR_KERNEL){
       // All functions must be marked as spir_func
       if(F.getCallingConv() != llvm::CallingConv::SPIR_FUNC)
@@ -165,9 +165,9 @@ bool LLVMToSpirvTranslator::toBackendFlavor(llvm::Module &M, PassHandler& PH) {
   // pointers. TODO: We should only remove them when we actually need to, and attempt
   // to fix them otherwise.
   llvm::SmallVector<llvm::CallBase*, 16> Calls;
-  for(auto& F : M.getFunctionList()) {
-    for(auto& BB : F.getBasicBlockList()) {
-      for(auto& I : BB.getInstList()) {
+  for(auto& F : M) {
+    for(auto& BB : F) {
+      for(auto& I : BB) {
         if(llvm::CallBase* CB = llvm::dyn_cast<llvm::CallBase>(&I)) {
           if (CB->getCalledFunction()->getName().startswith("llvm.lifetime.start") ||
               CB->getCalledFunction()->getName().startswith("llvm.lifetime.end")) {
@@ -287,9 +287,9 @@ bool LLVMToSpirvTranslator::optimizeFlavoredIR(llvm::Module& M, PassHandler& PH)
   // We adopt the workaround proposed there.
 
   llvm::SmallVector<llvm::Instruction*> InstsToRemove;
-  for(auto& F : M.getFunctionList()) {
-    for(auto& BB : F.getBasicBlockList()) {
-      for(auto& I : BB.getInstList()) {
+  for(auto& F : M) {
+    for(auto& BB : F) {
+      for(auto& I : BB) {
         if(auto* FI = llvm::dyn_cast<llvm::FreezeInst>(&I)) {
           FI->replaceAllUsesWith(FI->getOperand(0));
           FI->dropAllReferences();

--- a/src/compiler/sscp/HostKernelNameExtractionPass.cpp
+++ b/src/compiler/sscp/HostKernelNameExtractionPass.cpp
@@ -50,7 +50,7 @@ llvm::PreservedAnalyses HostKernelNameExtractionPass::run(llvm::Module &M,
 
   llvm::SmallVector<llvm::Function*> SSCPKernelNameExtractionFunctions;
 
-  for(auto& F : M.getFunctionList()) {
+  for(auto& F : M) {
     if(F.getName().find(SSCPExtractKernelNameIdentifier) != std::string::npos) {
       SSCPKernelNameExtractionFunctions.push_back(&F);
       for(auto U : F.users()) {

--- a/src/compiler/sscp/KernelOutliningPass.cpp
+++ b/src/compiler/sscp/KernelOutliningPass.cpp
@@ -113,8 +113,8 @@ public:
 
 private:
   void scanAllocas(llvm::Function *F, llvm::SmallDenseMap<llvm::Type *, int> &Scores) {
-    for(auto& BB : F->getBasicBlockList()) {
-      for(auto& I : BB.getInstList()) {
+    for(auto& BB : *F) {
+      for(auto& I : BB) {
         if(auto* AI = llvm::dyn_cast<llvm::AllocaInst>(&I)) {
           if(isKernelArgumentStruct(AI->getAllocatedType())) {
             Scores[AI->getAllocatedType()] = 0;
@@ -264,7 +264,7 @@ EntrypointPreparationPass::run(llvm::Module &M, llvm::ModuleAnalysisManager &AM)
         // however this cannot really happen as clang does not codegen our
         // attribute((annotate("hipsycl_sscp_outlining"))) for declarations
         // without definition.
-        if(F->getBasicBlockList().size() > 0)
+        if(F->size() > 0)
           this->OutliningEntrypoints.push_back(F->getName().str());
       }
     }
@@ -322,7 +322,7 @@ KernelOutliningPass::run(llvm::Module &M, llvm::ModuleAnalysisManager &AM) {
   }
   
   llvm::SmallVector<llvm::Function*, 16> PureHostFunctions;
-  for(auto& F: M.getFunctionList()) {
+  for(auto& F: M) {
     // Called Intrinsics don't show up in our device functions list,
     // so we need to treat them specially
     if(F.isIntrinsic()) {

--- a/src/compiler/sscp/TargetSeparationPass.cpp
+++ b/src/compiler/sscp/TargetSeparationPass.cpp
@@ -220,7 +220,7 @@ std::unique_ptr<llvm::Module> generateDeviceIR(llvm::Module &M,
   StringAttrsToRemove.push_back("target-cpu");
   StringAttrsToRemove.push_back("target-features");
   StringAttrsToRemove.push_back("tune-cpu");
-  for(auto& F : DeviceModule->getFunctionList()) {
+  for(auto& F : *DeviceModule) {
     for(auto& A : AttrsToRemove) {
       if(F.hasFnAttribute(A))
         F.removeFnAttr(A);
@@ -262,8 +262,8 @@ std::unique_ptr<llvm::Module> generateDeviceIR(llvm::Module &M,
 
    // Scan for imported function definitions
    ImportedSymbolsOutput.clear();
-  for(auto& F : DeviceModule->getFunctionList()) {
-    if(F.getBasicBlockList().size() == 0) {
+  for(auto& F : *DeviceModule) {
+    if(F.size() == 0) {
       // We currently use the heuristic that functions are imported
       // if they are not defined, not an intrinsic and don't start with
       // __ like our hipSYCL builtins. This is a hack, it would

